src/profiles-pycorelib/graph.py

```
@@ -0,0 +1,107 @@
+from profiles_rudderstack.material import WhtMaterial
+from profiles_rudderstack.logger import Logger
+from profiles_rudderstack.schema import MaterializationBuildSpecSchema
+from typing import List
+import pandas as pd
+from typing import Iterator
+import os
+import matplotlib
+matplotlib.use('Agg') 
+import matplotlib.pyplot as plt
+import threading
+
+
+class GraphModel(BaseModelType):
+    TypeName = "graph"
+    BuildSpecSchema = {
+        "type": "object",
+        "properties": {
+            "inputs": {"type": "array", "items": {"type": "string"}, "minItems": 1},
+            **MaterializationBuildSpecSchema["properties"],
+        },
+        "required": ["inputs"],
+        "additionalProperties": False
+    }
+
+    def __init__(self, build_spec: dict, schema_version: int, pb_version: str) -> None:
+        super().__init__(build_spec, schema_version, pb_version)
+
+    def get_material_recipe(self) -> PyNativeRecipe:
+        return GraphRecipe(self.build_spec["inputs"])
+
+    def validate(self):
+        return self.schema_version >= 51, "schema version should >= 51"
+
+
+class GraphRecipe(PyNativeRecipe):
+    def __init__(self, inputs: List[str]) -> None:
+        self.inputs = inputs
+        self.logger = Logger("graph_recipe")
+        self.sql = ""
+
+    def describe(self, this: WhtMaterial):
+        return self.sql, ".txt"
+
+    def prepare(self, this: WhtMaterial):
+        self.logger.info("Preparing")
+        for in_model in self.inputs:
+            this.de_ref(in_model)
+ 
+
+    def execute(self, this: WhtMaterial):
+        self.logger.info("Executing GraphModel ")
+        tables: List[pd.DataFrame] = []
+        output_folder=""
+        for in_model in self.inputs:
+            input_material = this.de_ref(in_model)
+            if input_material is None:
+                self.logger.info(
+                    "Input material for {0} not found".format(in_model))
+                continue
+            else:
```

No need for the else block as we have continue in if block, either remove the else block or the continue statement. Also because we are using `this.de_ref()` and not `this.de_ref` with `edge_type="optional"` if `input_material` is `None` we should raise an exception.

---

addressed.